File: packages/gbnf-experiment/data/20260909T180307Z_280f3890/ported_implementation/src/grammar-graph/get-serialized-rule-key.ts
import type { UnresolvedRule } from './grammar-graph-types.js';
import { isRuleChar, isRuleCharExclude, isRuleEnd, isRuleRef } from './type-guards.js';

export const KEY_TRANSLATION = {
  end: 0,
  char: 1,
  char_exclude: 2,
  ref: 3,
} as const;

export const getSerializedRuleKey = (rule: UnresolvedRule): string => {
  if (isRuleEnd(rule)) {
    return `${KEY_TRANSLATION.end}`;
  }

  if (isRuleChar(rule)) {
    return `${KEY_TRANSLATION.char}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleCharExclude(rule)) {
    return `${KEY_TRANSLATION.char_exclude}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleRef(rule)) {
    return `${KEY_TRANSLATION.ref}-${rule.value}`;
  }

  throw new Error(`Unknown rule type: ${JSON.stringify(rule)}`);
};

File: proposed-projects/round-trip-experiment/reverse/20260909T142951Z_4ff7cdec/20260910T082818Z_6d83d454/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key.py
from typing import Any

from .grammar_graph_types import RuleType
from .type_guards import is_rule_char, is_rule_char_exclude, is_rule_end, is_rule_ref

KEY_TRANSLATION = {
    RuleType.END: 0,
    RuleType.CHAR: 1,
    RuleType.CHAR_EXCLUDE: 2,
    RuleType.REF: 3,
}


def _serialize_value(value: Any) -> str:
    return "[" + ",".join(
        f"[{v[0]},{v[1]}]" if isinstance(v, (list, tuple)) else str(v) for v in value
    ) + "]"


def get_serialized_rule_key(rule: Any) -> str:
    if is_rule_end(rule):
        return f"{KEY_TRANSLATION[RuleType.END]}"

    if is_rule_char(rule):
        return f"{KEY_TRANSLATION[RuleType.CHAR]}-{_serialize_value(rule.value)}"

    if is_rule_char_exclude(rule):
        return (
            f"{KEY_TRANSLATION[RuleType.CHAR_EXCLUDE]}-{_serialize_value(rule.value)}"
        )

    if is_rule_ref(rule):
        return f"{KEY_TRANSLATION[RuleType.REF]}-{rule.value}"

    raise ValueError(f"Unknown rule type: {rule!r}")
